saul_coap.c

```diff
@@ -198,15 +204,86 @@ static ssize_t _saul_type_handler(coap_pkt_t* pdu, uint8_t *buf, size_t len, voi
         }
     }
 
-    /* TODO: Take care of all values. */
-    /* for (uint8_t i = 0; i < dim; i++) {
-       } */
+    CborError cbor_err = export_phydat_to_cbor(&encoder, cbor_buf, sizeof(cbor_buf), res, dim);
 
-    /* write the response buffer with the request device value */
-    resp_len += fmt_u16_dec((char *)pdu->payload, res.val[0]);
+    size_t buf_size = cbor_encoder_get_buffer_size(&encoder, cbor_buf);
+
+    if (cbor_err == CborNoError && buf_size > 0 && pdu->payload_len >= buf_size) {
+        memcpy(pdu->payload, cbor_buf, buf_size);
+        resp_len += buf_size;
+    } else {
+        resp_len = gcoap_response(pdu, buf, len, COAP_CODE_INTERNAL_SERVER_ERROR);
+    }
+
+    memset(cbor_buf, 0, sizeof(cbor_buf));
     return resp_len;
 }
 
+CborError export_phydat_to_cbor(CborEncoder *encoder, uint8_t *cbor_buf, size_t buf_len, phydat_t data, int dim)
+{
+    CborEncoder mapEncoder, aryEncoder;
+    CborError err = CborNoError;
+
+    cbor_encoder_init(encoder, cbor_buf, buf_len, 0);
+    if (err != CborNoError) {
+        return err;
+    }
+
+    err = cbor_encoder_create_map(encoder, &mapEncoder, 3);
```

Comment:
We should use `PHYDAT_DIM` instead of magic `3`.

Reply:
Nope. The dimension is not used in the map, but in the values array. And we don’t use `PHYDAY_DIM` there, but `dim` – the actual available dimensions.